src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -141,103 +141,97 @@ Error MeshForwarder::SendMessage(Message &aMessage)
 void MeshForwarder::HandleResolved(const Ip6::Address &aEid, Error aError)
 {
     Ip6::Address ip6Dst;
-    bool         enqueuedMessage = false;
+    bool         didUpdate = false;
 
-    for (Message &message : mResolvingQueue)
+    for (Message &message : mSendQueue)
     {
-        if (message.GetType() != Message::kTypeIp6)
+        if (!message.IsResolvingAddress())
         {
             continue;
         }
 
         IgnoreError(message.Read(Ip6::Header::kDestinationFieldOffset, ip6Dst));
 
-        if (ip6Dst == aEid)
+        if (ip6Dst != aEid)
         {
-            mResolvingQueue.Dequeue(message);
+            continue;
+        }
+
+        if (aError != kErrorNone)
+        {
+            LogMessage(kMessageDrop, message, kErrorAddressQuery);
+            mSendQueue.DequeueAndFree(message);
+            continue;
+        }
 
-            if (aError == kErrorNone)
-            {
 #if OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
-                // Pass back to IPv6 layer for DUA destination resolved by Backbone Query
-                if (ForwardDuaToBackboneLink(message, ip6Dst) != kErrorNone)
-#endif
-                {
-                    mSendQueue.Enqueue(message);
-                    enqueuedMessage = true;
-                }
-            }
-            else
-            {
-                LogMessage(kMessageDrop, message, aError);
-                message.Free();
-            }
+        // Pass back to IPv6 layer for DUA destination resolved
+        // by Backbone Query
+        if (ShouldForwardDuaToBackboneLink(message, ip6Dst))
+        {
+            mSendQueue.Dequeue(message);
+            IgnoreError(Get<Ip6::Ip6>().HandleDatagram(message, nullptr, nullptr, /* aFromHost */ false));
+            continue;
         }
+#endif
+
+        message.SetResolvingAddress(false);
+        didUpdate = true;
     }
 
-    if (enqueuedMessage)
+    if (didUpdate)
     {
         mScheduleTransmissionTask.Post();
     }
 }
 
 #if OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
-Error MeshForwarder::ForwardDuaToBackboneLink(Message &aMessage, const Ip6::Address &aDst)
+bool MeshForwarder::ShouldForwardDuaToBackboneLink(Message &aMessage, const Ip6::Address &aDst)
 {
-    Error   error = kErrorNone;
+    bool    shouldForward = false;
     uint8_t ttl;
 
-    VerifyOrExit(Get<BackboneRouter::Local>().IsPrimary() && Get<BackboneRouter::Leader>().IsDomainUnicast(aDst),
-                 error = kErrorNoRoute);
+    VerifyOrExit(Get<BackboneRouter::Local>().IsPrimary() && Get<BackboneRouter::Leader>().IsDomainUnicast(aDst));
+
+    VerifyOrExit(Get<AddressResolver>().LookUp(aDst) == Get<Mle::MleRouter>().GetRloc16());
 
-    VerifyOrExit(Get<AddressResolver>().LookUp(aDst) == Get<Mle::MleRouter>().GetRloc16(), error = kErrorNoRoute);
+    shouldForward = true;
 
     // Avoid decreasing TTL twice
     IgnoreError(aMessage.Read(Ip6::Header::kHopLimitFieldOffset, ttl));
```

Comment:
`Should...` methods usually don't modify any data, but the TTL is still changed in this method. 
This might be a surprise for code readers?

Reply:
Good point. I agree. Will move the TTL change to the method calling this.

Reply:
Changed this in new push (added it as sperate commit over the existing one to make it easier to check and review).
It actually removes the `ShouldForward..` method and have the check done inline in `HandleResolved()`.
@simonlingoogle PTAL.